src/pkg/retention/launcher_test.go

```diff
@@ -0,0 +1,218 @@
+	_ "github.com/goharbor/harbor/src/pkg/retention/res/selectors/regexp"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type fakeProjectManager struct {
+	projects []*models.Project
+}
+
+func (f *fakeProjectManager) List(...*models.ProjectQueryParam) ([]*models.Project, error) {
+	return f.projects, nil
+}
+func (f *fakeProjectManager) Get(idOrName interface{}) (*models.Project, error) {
+	id, ok := idOrName.(int64)
+	if ok {
+		for _, project := range f.projects {
+			if project.ProjectID == id {
+				return project, nil
+			}
+		}
+		return nil, nil
+	}
+	name, ok := idOrName.(string)
+	if ok {
+		for _, project := range f.projects {
+			if project.Name == name {
+				return project, nil
+			}
+		}
+		return nil, nil
+	}
+	return nil, fmt.Errorf("invalid parameter: %v, should be ID(int64) or name(string)", idOrName)
+}
+
+type fakeRepositoryManager struct {
+	imageRepositories []*models.RepoRecord
+	chartRepositories []*chartserver.ChartInfo
+}
+
+func (f *fakeRepositoryManager) ListImageRepositories(projectID int64) ([]*models.RepoRecord, error) {
+	return f.imageRepositories, nil
+}
+func (f *fakeRepositoryManager) ListChartRepositories(projectID int64) ([]*chartserver.ChartInfo, error) {
+	return f.chartRepositories, nil
+}
+
+type fakeClient struct {
+	id int
+}
+
+func (f *fakeClient) GetCandidates(repo *res.Repository) ([]*res.Candidate, error) {
+	return nil, nil
+}
+
+func (f *fakeClient) Delete(candidate *res.Candidate) error {
+	return nil
+}
+func (f *fakeClient) SubmitTask(repository *res.Repository, meta *policy.LiteMeta) (string, error) {
+	f.id++
+	return strconv.Itoa(f.id), nil
+}
+
+func TestGetProjects(t *testing.T) {
+	pro := &models.Project{
+		ProjectID: 1,
+		Name:      "library",
+	}
+	project.Mgr = &fakeProjectManager{
+		projects: []*models.Project{
+			pro,
+		}}
+	projects, err := getProjects()
+	require.Nil(t, err)
+	assert.Equal(t, 1, len(projects))
+	assert.Equal(t, pro.ProjectID, projects[0].NamespaceID)
+	assert.Equal(t, pro.Name, projects[0].Namespace)
+}
+
+func TestGetRepositories(t *testing.T) {
+	pro := &models.Project{
+		ProjectID: 1,
+		Name:      "library",
+	}
+	project.Mgr = &fakeProjectManager{
+		projects: []*models.Project{
+			pro,
+		}}
+	repository.Mgr = &fakeRepositoryManager{
+		imageRepositories: []*models.RepoRecord{
+			{
+				Name: "library/image",
+			},
+		},
+		chartRepositories: []*chartserver.ChartInfo{
+			{
+				Name: "chart",
+			},
+		},
+	}
+
+	repositories, err := getRepositories(1)
+	require.Nil(t, err)
+	assert.Equal(t, 2, len(repositories))
+	assert.Equal(t, "library", repositories[0].Namespace)
+	assert.Equal(t, "image", repositories[0].Repository)
+	assert.Equal(t, "image", repositories[0].Kind)
+	assert.Equal(t, "library", repositories[1].Namespace)
+	assert.Equal(t, "chart", repositories[1].Repository)
+	assert.Equal(t, "chart", repositories[1].Kind)
+}
+
+func TestLaunch(t *testing.T) {
```

Comment:
TODO:

use `github.com/stretchr/testify` drive UT cases. Please refine the UT in next PR